pkg/backup.go

```diff
@@ -192,41 +203,148 @@ func (opt *vaultOptions) backupVault(targetRef api_v1beta1.TargetRef) (*restic.B
 	if err != nil {
 		return nil, err
 	}
 
-	session.setUserArgs(opt.vaultArgs)
+	// set the vault token that has the necessary permission to save or restore snapshot
+	err = session.setVaultToken(opt.kubeClient, appBinding, vs)
+	if err != nil {
+		return nil, err
+	}
+
+	// set the vault connection parameters, essentially the vault leader node address
+	err = session.setVaultConnectionParameters(vaultClient, appBinding)
+	if err != nil {
+		return nil, err
+	}
+
+	klog.Infof("Trying to backup for VaultServer %s/%s\n", vs.Namespace, vs.Name)
+
+	// save the vault snapshot in the interim directory running command like: vault operator raft snapshot save backup.snap
+	err = opt.saveVaultSnapshot(session)
+	if err != nil {
+		return nil, err
+	}
+
+	// save the unseal keys & root token too, for complete backup
+	// write these in the interim directory along with the snapshot file
+	if err := opt.writeVaultTokenKeys(vs); err != nil {
+		return nil, err
+	}
 
-	// add backup command in the pipeline
-	opt.backupOptions.StdinPipeCommands = append(opt.backupOptions.StdinPipeCommands, *session.cmd)
-	resticWrapper, err := restic.NewResticWrapperFromShell(opt.setupOptions, session.sh)
+	// now take the backup of this directory using stash/restic
+	opt.backupOptions.BackupPaths = []string{opt.interimDataDir}
+	resticWrapper, err := restic.NewResticWrapper(opt.setupOptions)
 	if err != nil {
 		return nil, err
 	}
 
 	return resticWrapper.RunBackup(opt.backupOptions, targetRef)
 }
+
+func (opt *VaultOptions) saveVaultSnapshot(session *sessionWrapper) error {
+	klog.Infoln("Trying to save snapshot")
+	session.cmd.Args = append(session.cmd.Args, "operator", "raft", "snapshot", "save", filepath.Join(opt.interimDataDir, VaultSnapshotFile))
+
+	session.sh.ShowCMD = false
+	session.setUserArgs(opt.vaultArgs)
+	session.sh.Command(VaultCMD, session.cmd.Args...)
+
+	if err := session.sh.Run(); err != nil {
+		return err
+	}
+
+	klog.Infoln("snapshot saved successfully")
+	return nil
+}
+
+func (opt *VaultOptions) writeVaultTokenKeys(vs *vaultapi.VaultServer) error {
+	klog.Infoln("Trying to get, write unseal keys & root token")
+	keyPrefix, err := opt.getKeyPrefix()
+	if err != nil {
+		return err
+	}
+	opt.keyPrefix = keyPrefix
+
+	// create a new store interface
+	// for backup:
+	// i. Get the unseal key & root token from store based on the unseal mode
+	// ii. write them into the interim directory which will be backed up
+	st, err := store.NewStore(opt.kubeClient, vs)
+	if err != nil {
+		return err
+	}
+
+	var keys []string
+	keys = append(keys, opt.tokenName())
+	for i := 0; i < int(vs.Spec.Unsealer.SecretShares); i++ {
+		keys = append(keys, opt.unsealKeyName(i))
+	}
+
+	for _, key := range keys {
+		value, err := st.Get(key)
+		if err != nil {
+			klog.Errorf("failed to get key %s with %s\n", key, err.Error())
+			return err
+		}
+
+		if err := opt.write(key, value); err != nil {
+			klog.Errorf("failed to write key %s with %s\n", key, err.Error())
+		}
```

Comment:
will it affect the restore process if a key is not written? If yes. Should we return the error from here?